### PR TITLE
config: add support for host header rewriting

### PR DESCRIPTION
**Pull request by calebdoxsey**
## Summary
This PR adds new policy config options for customizing host header rewriting. I basically copied the available options in envoy: https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto.html#config-route-v3-routeaction.

I updated the readme with an explanation of the value:

### Host Rewrite

- `yaml`/`json` settings: `host_rewrite`, `host_rewrite_header`, `host_path_regex_rewrite_pattern`, `host_path_regex_rewrite_substitution`
- Type: `string`
- Optional
- Example: `host_rewrite: "example.com"`

The `host` header can be customized via 3 mutually exclusive options:

1. `host_rewrite` which will rewrite the host to a new literal value.
2. `host_rewrite_header` which will rewrite the host to match an incoming header value.
3. `host_path_regex_rewrite_pattern`, `host_path_regex_rewrite_substitution` which will rewrite the host according to a regex matching the path. For example with the following config:

    ```yaml
    host_path_regex_rewrite_pattern: "^/(.+)/.+$"
    host_path_regex_rewrite_substitution: \1
    ```

    Would rewrite the host header to `example.com` given the path `/example.com/some/path`.

These options correspond to the envoy route action host related options, which can be found [here](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/route/v3/route_components.proto.html#config-route-v3-routeaction).

## Related issues
- Implements #1452 

**Checklist**:
- [x] add related issues
- [x] updated docs
- [x] updated unit tests
- [ ] updated UPGRADING.md
- [x] ready for review

### CODE REVIEW

**Comment by calebdoxsey** [comment]
@desimone @travisgroth here's an implementation of the host header rewriting. I added all the possible options I could find in envoy, but we could remove some (or rename them) if we don't think they're useful.

We already had one option (`preserve_host_header`), so these changes should also handle that.

**Review by desimone — COMMENTED** [review]
Code LGTM. Nice work. 

I think it makes sense to include all the rewrite options that envoy provides; and can't think of a good reason not to.  @travisgroth thoughts?

**Comment by travisgroth** [comment]
> I think it makes sense to include all the rewrite options that envoy provides; and can't think of a good reason not to. @travisgroth thoughts?

Agreed.  Seems pretty low overhead, and is better than handling them one at a time as use cases pop up. 👍

File: docs/reference/readme.md [hunk 1]
```diff
@@ -1057,6 +1057,28 @@ prefix_rewrite: /
...
+1. `host_rewrite` which will rewrite the host to a new literal value.
+2. `host_rewrite_header` which will rewrite the host to match an incoming header value.
+3. `host_path_regex_rewrite_pattern`, `host_path_regex_rewrite_substitution` which will rewrite the host according to a regex matching the path. For example with the following config:
+
+    ```yaml
+    host_path_regex_rewrite_pattern: "^/(.+)/.+$"
+    host_path_regex_rewrite_substitution: \1
+    ```
+
+    Would rewrite the host header to `example.com` given the path `/example.com/some/path`.
```
Comment (desimone) [inline docs/reference/readme.md:1078]:
```suggestion
Would rewrite the host header to `example.com` given the path `/example.com/some/path`.
```

**Review by desimone — APPROVED** [review]

**Pull request merged**